db/view/view.cc

```diff
@@ -1754,6 +1754,36 @@ future<> view_update_generator::mutate_MV(
     });
 }
 
+
+utils::small_vector<gms::inet_address, 2> view_update_generator::get_endpoints_for_view_update(
+        dht::token base_token,
+        const frozen_mutation_and_schema& view_update) const {
+    dht::token view_token = dht::get_token(*view_update.s, view_update.fm.key());
+    const sstring& keyspace_name = view_update.s->ks_name();
+
+    // For every base table replica there's a single paired view replica.
+    std::optional<gms::inet_address> paired_replica_endpoint =
+        get_view_natural_endpoint(_proxy.local().local_db(), keyspace_name, base_token, view_token);
+
+    // In case of an ongoing topology change it might also be necessary to send the update to pending nodes.
+    inet_address_vector_topology_change pending_endpoints =
+        _proxy.local().get_token_metadata_ptr()->pending_endpoints_for(view_token, keyspace_name);
```

Comment:
The function should get an effective_replication_map_ptr from its caller and use that. The caller should keep the effective_replication_map_ptr captured until it stops using the results of this calculation.

The reason is that the calculation can be outdated as soon as the function returns. Capturing the effective_replication_map_ptr keeps track of how many users reference the older topology, and later we'll use it as a barrier to prevent the topology state machine from moving forward.

/cc @bhalevy

Reply:
I read up a bit about topology changes (mostly just https://github.com/scylladb/scylladb/blob/master/docs/dev/topology-over-raft.md), AFAIU the gist of it is:

When a new node is added it takes over some token ranges. While it's joining the cluster we send the updates to both the old owner of the token range and the new node. Once the joining is finished we switch over to the new node completely.

There is a risk of the following race:
* We have a view update to send, from the current topology we deduce it should be sent to node A
* While the update is being sent a topology change happens and now it should be sent to node B instead
* The update is sent to the wrong node, view inconsistency happens

To avoid this each update should get an instance of `effective_replication_map_ptr` which prevents the switch to B until the change has fully completed, and it should keep this `ptr` until the update has completed fully.

If this understanding is correct, then the current view update sending is buggy, as it doesn't use the `effective_replication_map_ptr` at all:
https://github.com/scylladb/scylladb/blob/20f620feb9a26ca0d7a78018fc79ac522a52513a/db/view/view.cc#L1639-L1640

I guess it should be changed, although I'm not sure if it belongs in this PR.

Reply:
Regarding the new code - maybe the throttling calculation doesn't really need to lock the topology? It's just a heuristic, information about remote backlogs isn't 100% up to date and the calculation doesn't need to be 100% exact, it's just an approximation of how much work the other nodes are doing.

There's a risk that comes with locking the topology for the whole duration of a throttled operation. We do a sleep that could last arbitrarily long, possibly even many seconds. In the usual case the topology is only locked for the amount of time required to send a single request, but here the lock might be a lot longer. Can the topology code handle being locked for a whole minute? Or will timeouts start popping up?

I guess it's also an issue with normal view updates, even without the throttling. If we lock the topology at the point of generating a view update, then the lock will be kept for the amount of time that the update is in the queue. The queue of view updates can take up 10% of the available memory, so there could be a significant delay between generating the view update and actually sending it. This could be fixed by locking the topology at the moment of sending, but that would require a bit of refactoring.

Overall I think this PR would be better off without any topology locking - taking the lock before sleeping could cause issues, and the throttling calculation doesn't need to be 100% exact, so IMO the pros of not locking outweights the cons.